xdsl/parser/affine_parser.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+from xdsl.parser.core import Parser, ParserState, Token, ParseError
+from xdsl.ir.affine import AffineExpr, AffineMap
+
+
+class AffineParser(Parser):
+    _BINOP_PRECEDENCE = {
+        "+": 10,
+        "-": 10,
+        "*": 20,
+        "ceildiv": 20,
+        "floordiv": 20,
+        "mod": 20,
+    }
+
+    def __init__(self, state: ParserState) -> None:
+        self.resume_from_state(state)
+
+    def _parse_primary(self, dims: list[str], syms: list[str]) -> AffineExpr:
+        """
+        primary ::= `(` affine-expr `)`
+                  | bare-id
+                  | integer-literal
+                  | `-` primary
+        """
+        # Handle parentheses
+        if self._parse_optional_token(Token.Kind.L_PAREN):
+            expr = self._parse_affine_expr(dims, syms)
+            self._consume_token(Token.Kind.R_PAREN)
```

Comment:
```suggestion
            self._parse_token(Token.Kind.R_PAREN)
```
`_consume_token` expects that you already know the kind of token you are parsing, and triggers an assertion.
`_parse_token` will parse the token, and raise an error if it is not the right kind.